Wrappers/Python/cil/optimisation/utilities/sampler.py

```diff
@@ -0,0 +1,629 @@
+# -*- coding: utf-8 -*-
+#   This work is part of the Core Imaging Library (CIL) developed by CCPi
+#   (Collaborative Computational Project in Tomographic Imaging), with
+#   substantial contributions by UKRI-STFC and University of Manchester.
+
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+
+#   http://www.apache.org/licenses/LICENSE-2.0
+
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+
+import numpy as np
+import math
+import time
+from functools import partial
+
+
+class Sampler():
+    """     
+    The user is recommended to not instantiate this class  directly but instead use one of the static methods that will return instances of different samplers.
+    
+    This class wraps a function that takes an in integer iteration number and returns an integer from {0, 1, …, S-1} with S=num_indices to produce a Sampler which returns a new index for each call of `next`.
+
+    
+    Custom samplers can be created by subclassing this sampler class. 
+
+    
+
+    Parameters
+    ----------
+
+    function: A function that takes an in integer iteration number and returns an integer from {0, 1, …, S-1} with S=num_indices. 
+
+    num_indices: int
+        One above the largest integer that could be drawn by the sampler. The sampler will select from a list of indices {0, 1, …, S-1} with S=num_indices. 
+
+    sampling_type:str default is None 
+        The sampling type used. Choose from  "sequential", "staggered", "herman_meyer" and "from_function". 
+
+    prob_weights: list of floats of length num_indices that sum to 1.  Default is [1/num_indices]*num_indices 
+        Consider that the sampler is called a large number of times this argument holds the expected number of times each index would be called,  normalised to 1. 
+
+    Returns
+    -------
+    A Sampler that can be called with Sampler.next() or next(Sampler)
```

Comment:
returns should have the format:
```
Returns
--------
type
    description
````

So here: 
Sampler
        An instance of the Sampler class representing the desired configuration.